README.md
Bluetree Symfony Console Styles Extender
========================================

[![Latest Stable Version](https://poser.pugx.org/bluetree-service/symfony-console-style/v/stable.svg?style=flat-square)](https://packagist.org/packages/bluetree-service/symfony-console-style)
[![Total Downloads](https://poser.pugx.org/bluetree-service/symfony-console-style/downloads.svg?style=flat-square)](https://packagist.org/packages/bluetree-service/symfony-console-style)
[![License](https://poser.pugx.org/bluetree-service/symfony-console-style/license.svg?style=flat-square)](https://packagist.org/packages/bluetree-service/symfony-console-style)

##### Builds
| Travis | Scrutinizer | Circle-CI |
|:---:|:---:|:---:|
| [![Build Status](https://travis-ci.org/bluetree-service/symfony-console-style.svg?style=flat-square)](https://travis-ci.org/bluetree-service/symfony-console-style) | [![Build Status](https://scrutinizer-ci.com/g/bluetree-service/symfony-console-style/badges/build.png?style=flat-square&b=master)](https://scrutinizer-ci.com/g/bluetree-service/symfony-console-style/build-status/master) | [![<Bluetree Service>](https://circleci.com/gh/bluetree-service/symfony-console-style.svg?style=shield)](https://app.circleci.com/pipelines/github/bluetree-service/symfony-console-style) |

##### Coverage
| Coveralls | Scrutinizer |
|:---:|:---:|
| [![Coverage Status](https://coveralls.io/repos/github/bluetree-service/symfony-console-style/badge.svg?style=flat-square&branch=master)](https://coveralls.io/github/bluetree-service/symfony-console-style?branch=master) | [![Code Coverage](https://scrutinizer-ci.com/g/bluetree-service/symfony-console-style/badges/coverage.png?style=flat-square&b=master)](https://scrutinizer-ci.com/g/bluetree-service/symfony-console-style/?branch=master) |

##### Quality
| Code Climate | Scrutinizer | Sensio Labs |
|:---:|:---:|:---:|
| [![Code Climate](https://codeclimate.com/github/bluetree-service/symfony-console-style/badges/gpa.svg?style=flat-square)](https://codeclimate.com/github/bluetree-service/symfony-console-style) | [![Scrutinizer Code Quality](https://scrutinizer-ci.com/g/bluetree-service/symfony-console-style/badges/quality-score.png?style=flat-square&b=master)](https://scrutinizer-ci.com/g/bluetree-service/symfony-console-style/?branch=master) | [![SensioLabsInsight](https://insight.sensiolabs.com/projects/e9b6307b-4407-4fbe-8e40-233a3ec7f352/mini.png?style=flat-square)](https://insight.sensiolabs.com/projects/e9b6307b-4407-4fbe-8e40-233a3ec7f352) |
|  | [![Code Intelligence Status](https://scrutinizer-ci.com/g/bluetree-service/symfony-console-style/badges/code-intelligence.svg?style=flat-square&b=master)](https://scrutinizer-ci.com/code-intelligence) |  |

Extended Symfony Console styles. Including new blocks, message time, interactive checkboxes and many others. 

### Included libraries
* **BlueStyle\MultiSelect** - Handle single & multi select
* **BlueStyle\Style** - Nes styles

Documentation & Examples
------------------------
* [Multi select](https://github.com/bluetree-service/symfony-console-style/tree/examples/multiselect.php "Multi select")
* [Single select](https://github.com/bluetree-service/symfony-console-style/tree/develop/singleselect.php "Single select")
* [Extended styles](https://github.com/bluetree-service/symfony-console-style/tree/develop/style.php "Extended styles")


Install via Composer
--------------
To use packages you can just download package and pace it in your code. But recommended
way to use _BlueConsole_ is install it via Composer. To include _BlueConsole_
libraries paste into composer json:

```json
{
    "require": {
        "bluetree-service/symfony-console-style": "version_number"
    }
}
```

Project description
--------------

### Used conventions

* **Namespaces** - each library use namespaces
* **PSR-2** - [PSR-2](http://www.php-fig.org/psr/psr-2/) coding standard
* **PSR-4** - [PSR-4](http://www.php-fig.org/psr/psr-4/) auto loading standard
* **Composer** - [Composer](https://getcomposer.org/)

### Requirements

* PHP 7.2 or higher

Change log
--------------
All release version changes:  
[Change log](https://github.com/bluetree-service/symfony-console-style/tree/develop/doc/changelog.md "Change log")

License
--------------
This bundle is released under the Apache license.  
[Apache license](https://github.com/bluetree-service/symfony-console-style/LICENSE "Apache license")

Travis Information
--------------
[Travis CI Build Info](https://travis-ci.org/bluetree-service/symfony-console-style)


doc/CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

## 0.6.1.1 - 2022-07-03
### Added
* Circle-CI badge
### Changed
* Fixed composer.json version

## 0.6.1.0 - 2022-05-22
### Changed
* List rendering timeout changed from 0.5s to 0.1s

**## 0.6.0.0 - 2020-08-09
### Added
* An added helper to run commands in multiple processes
* Styles for multiple process progress bar
* Floating progress bar
* Multi line block
* Circle CI configuration
* README description
### Changed
* PSR-12 fixes
* Fixes from CI/CD services**

## 0.5.0.0 - 2020-07-05
### Added
* Message timer with date time
* Style tests
### Changed
* Style::renderBlock make public
* PSR-12 fixes

## 0.4.0.0 - 2019-05-04
### Added
* Message timer
### Changed
* Some block styles
* Split `Style` class

## 0.3.0.4 - 2018-12-18
### Added
* Some new tests
* Some todo in `src/Style.php`
* New examples
* Coverage composer script

## 0.3.0.3 - 2018-11-11
### Added
* Clover reporter as dev dependency

## 0.3.0.2 - 2018-11-06
### Removed
* Support for PHP 7.0

## 0.3.0.1 - 2018-10-31
### Changed
* Symfony console required version changed from 4.* to ~4.0

## 0.3.0.0 - 2018-10-23
### Added
* Some example files (./examples)
* Some additional styles
* Symfony var-dumper
* Basic animation for process waiting class
### Changed
* Symfony console required version changed from 4.0.* to 4.*
* Fixed errorLine with multiple messages
* Fixed single select
### Removed
* PHP 5.6 support
* composer.lock as library convention approach

## 0.2.0.0 - 2018-09-14
### Added
* Basic display style
### Changed
* Minimal PHP requirements changed to 7.0

## 0.1.0.0 - 2016-01-29
### Added
* Proof of concept
* Composer dependencies


tests/StyleTest.php
<?php

namespace Test;

use PHPUnit\Framework\TestCase;
use Symfony\Component\Console\Input\ArgvInput;
use Symfony\Component\Console\Output\StreamOutput;
use Symfony\Component\Console\Helper\FormatterHelper;
use Symfony\Component\Console\Tester\TesterTrait;
use BlueConsole\Style;

class StyleTest extends TestCase
{
    use TesterTrait;

    /**
     * @var Style
     */
    protected $style;

    public function setUp(): void
    {
        $input = new ArgvInput();
        $this->output = new StreamOutput(fopen('php://memory', 'w+', false));
        $formatter = new FormatterHelper();
        $this->style = new Style($input, $this->output, $formatter);
    }

    public function testTimer(): void
    {
        $this->style->timer();
        $this->assertRegExp('/\[ 000000000.00[\d]{2} \]/', $this->getDisplay());

        $this->style->timer(true);
        $this->assertRegExp('/\[ 000000000.00[\d]{2} \]\n/', $this->getDisplay());
    }

    public function testFormatSection(): void
    {
        $this->style->formatSection('Section', 'Format Section info (default)');
        $this->assertEquals("[Section] Format Section info (default)\n", $this->getDisplay());
    }

    public function testFormatBlockArray(): void
    {
        $expect = <<<EOT
<ok>      </ok>
<ok>  m1  </ok>
<ok>      </ok>
<ok>      </ok>
<ok>  m2  </ok>
<ok>      </ok>

EOT;

        $this->style->formatBlock(['m1', 'm2'], 'ok', true);
        $this->assertEquals($expect, $this->getDisplay());
    }

    public function testFormatBlock(): void
    {
        $expect = <<<EOT
<ok>      </ok>
<ok>  m1  </ok>
<ok>      </ok>

EOT;

        $this->style->formatBlock('m1', 'ok', true);
        $this->assertEquals($expect, $this->getDisplay());
    }

    public function testErrorLine(): void
    {
        $this->style->errorLine(['m1']);
        $this->assertEquals(" m1 \n", $this->getDisplay());
    }

    public function testGenericBlock(): void
    {
        $expected = <<<EOT
                                                   
  [ERROR] m1                                       
                                                   


EOT;

        $this->style->genericBlock('m1', 'red', 'error', 50);
        $this->assertEquals($expected, $this->getDisplay());
    }

    public function testGenericBlockWithMissingType(): void
    {
        $this->style->genericBlock('m1', 'red', '');
        $this->assertEquals('', $this->getDisplay());
    }

    public function testNote(): void
    {
        $expected = <<<EOT
                                                                                                     
  [NOTE] m1                                                                                          
                                                                                                     


EOT;

        $this->style->note('m1');
        $this->assertEquals($expected, $this->getDisplay());
    }

    public function testCaution(): void
    {
        $expected = <<<EOT
                                                                                                     
  [CAUTION] m1                                                                                       
                                                                                                     


EOT;

        $this->style->caution('m1');
        $this->assertEquals($expected, $this->getDisplay());
    }

    public function testSuccess(): void
    {
        $expected = <<<EOT
                                                                                                     
  [SUCCESS] m1                                                                                       
                                                                                                     


EOT;

        $this->style->success('m1');
        $this->assertEquals($expected, $this->getDisplay());
    }

    public function testWarning(): void
    {
        $expected = <<<EOT
                                                                                                     
  [WARNING] m1                                                                                       
                                                                                                     


EOT;

        $this->style->warning('m1');
        $this->assertEquals($expected, $this->getDisplay());
    }

    public function testError(): void
    {
        $expected = <<<EOT
                                                                                                     
  [ERROR] m1                                                                                         
                                                                                                     


EOT;

        $this->style->error('m1');
        $this->assertEquals($expected, $this->getDisplay());
    }
}


tests/AdditionalTest.php
<?php

namespace Test;

use PHPUnit\Framework\TestCase;
use Symfony\Component\Console\Input\ArgvInput;
use Symfony\Component\Console\Output\StreamOutput;
use Symfony\Component\Console\Helper\FormatterHelper;
use Symfony\Component\Console\Tester\TesterTrait;
use BlueConsole\Style;

class AdditionalTest extends TestCase
{
    use TesterTrait;

    /**
     * @var Style
     */
    protected $style;

    public function setUp(): void
    {
        $input = new ArgvInput();
        $this->output = new StreamOutput(\fopen('php://memory', 'wb+', false));
        $formatter = new FormatterHelper();
        $this->style = new Style($input, $this->output, $formatter);
    }

    public function testTruncateMessage(): void
    {
        $message = $this->style->truncateMessage('1234567890', 5);
        $this->assertEquals('12345...', $message);

        $message = $this->style->truncateMessage('1234567890', 5, '-');
        $this->assertEquals('12345-', $message);
    }

    public function testTruncate(): void
    {
        $this->style->truncate('1234567890', 5);
        $this->assertEquals('12345...', $this->getDisplay());
    }

    public function testTruncateWithSuffix(): void
    {
        $this->style->truncate('1234567890', 5, '_');
        $this->assertEquals('12345_', $this->getDisplay());
    }

    public function testTruncateLn(): void
    {
        $this->style->truncateLn('1234567890', 5);
        $this->assertEquals("12345...\n", $this->getDisplay());
    }

    public function testAlignGrtSet(): void
    {
        $this->assertEquals(12, $this->style->getAlign());
        $this->assertInstanceOf(Style::class, $this->style->setAlign(100));
        $this->assertEquals(100, $this->style->getAlign());
    }

    /**
     * @dataProvider alignDataProvider
     */
    public function testAlign($string, $expect): void
    {
        $this->assertEquals($expect, $this->style->align($string, 15));
    }

    public function alignDataProvider(): array
    {
        return [
            [
                'string' => 5,
                'expect' => '           ',
            ],
            [
                'string' => '5',
                'expect' => '               ',
            ],
            [
                'string' => ['123', '456'],
                'expect' => '        ',
            ],
            [
                'string' => null,
                'expect' => '                ',
            ],
        ];
    }
}


tests/StyleMessageTest.php
<?php

namespace Test;

use PHPUnit\Framework\TestCase;
use Symfony\Component\Console\Input\ArgvInput;
use Symfony\Component\Console\Output\StreamOutput;
use Symfony\Component\Console\Helper\FormatterHelper;
use Symfony\Component\Console\Tester\TesterTrait;
use BlueConsole\Style;

class StyleMessageTest extends TestCase
{
    use TesterTrait;

    /**
     * @var Style
     */
    protected $style;

    public function setUp(): void
    {
        $input = new ArgvInput;
        $this->output = new StreamOutput(fopen('php://memory', 'w+', false));
        $formatter = new FormatterHelper;
        $this->style = new Style($input, $this->output, $formatter);
    }

    public function testWarningMessage(): void
    {
        $this->style->warningMessage('Warning message');
        $this->assertEquals($this->getDisplay(), "[ WARN ]     Warning message\n");
    }

    public function testErrorMessage(): void
    {
        $this->style->errorMessage('Error message');
        $this->assertEquals($this->getDisplay(), "[ FAIL ]     Error message\n");
    }

    public function testOkMessage(): void
    {
        $this->style->okMessage('Ok message');
        $this->assertEquals($this->getDisplay(), "[  OK  ]     Ok message\n");
    }

    public function testInfoMessage(): void
    {
        $this->style->infoMessage('Info message');
        $this->assertEquals($this->getDisplay(), "[ INFO ]     Info message\n");
    }

    public function testToggleTimer(): void
    {
        $this->style->toggleShowTimer();

        $this->assertTrue($this->style->isTimerOn());
    }

    public function testDateFormatting(): void
    {
        $this->assertEquals('c', $this->style->getDateTimeFormat());
        $this->assertNull($this->style->setDateTimeFormat('d-m-Y'));
        $this->assertEquals('d-m-Y', $this->style->getDateTimeFormat());

        $this->assertEquals(14, $this->style->getTimeCharLength());
        $this->assertInstanceOf(Style::class, $this->style->setTimeCharLength(1));
        $this->assertEquals(1, $this->style->getTimeCharLength());
    }

    public function testWarningMessageWithTimer(): void
    {
        $this->style->toggleShowTimer();
        $this->style->warningMessage('Warning message');

        $this->assertRegExp($this->getMessageExpression('warning'), $this->getDisplay());
    }

    public function testOkMessageWithTimer(): void
    {
        $this->style->toggleShowTimer();
        $this->style->okMessage('Ok message');

        $this->assertRegExp($this->getMessageExpression('ok'), $this->getDisplay());
    }

    public function testErrorMessageWithTimer(): void
    {
        $this->style->toggleShowTimer();
        $this->style->errorMessage('Error message');

        $this->assertRegExp($this->getMessageExpression('error'), $this->getDisplay());
    }

    public function testInfoMessageWithTimer(): void
    {
        $this->style->toggleShowTimer();
        $this->style->infoMessage('Info message');

        $this->assertRegExp($this->getMessageExpression('info'), $this->getDisplay());
    }

    public function testWarningMessageWithDateTime(): void
    {
        $this->style->toggleShowTimer();
        $this->style->toggleTimerType();
        $this->style->warningMessage('Warning message');

        $this->assertRegExp($this->getMessageExpression('warning', true), $this->getDisplay());
    }

    public function testOkMessageWithDateTime(): void
    {
        $this->style->toggleShowTimer();
        $this->style->toggleTimerType();
        $this->style->okMessage('Ok message');

        $this->assertRegExp($this->getMessageExpression('ok', true), $this->getDisplay());
    }

    public function testErrorMessageWithDateTime(): void
    {
        $this->style->toggleShowTimer();
        $this->style->toggleTimerType();
        $this->style->errorMessage('Error message');

        $this->assertRegExp($this->getMessageExpression('error', true), $this->getDisplay());
    }

    public function testInfoMessageWithDateTime(): void
    {
        $this->style->toggleShowTimer();
        $this->style->toggleTimerType();
        $this->style->infoMessage('Info message');

        $this->assertRegExp($this->getMessageExpression('info', true), $this->getDisplay());
    }

    /**
     * @param string $type
     * @param bool $isDateTime
     * @return string
     */
    protected function getMessageExpression(string $type, bool $isDateTime = false): string
    {
        $begin = '/\[ 000000000.00[\d]{2} \]\[ ';

        if ($isDateTime) {
            $begin = '/\[ 2[\d]{3}-[\d]{2}-[\d]{2}T[\d]{2}:[\d]{2}:[\d]{2}\+[\d]{2}:[\d]{2} \]\[ ';
        }

        $middle = ' \][ ]{5}';
        $end = ' message\\n/';

        switch ($type) {
            case 'warning':
                return $begin . 'WARN' . $middle . 'Warning' . $end;
            case 'ok':
                return $begin . ' OK ' . $middle . 'Ok' . $end;
            case 'error':
                return $begin . 'FAIL' . $middle . 'Error' . $end;
            case 'info':
                return $begin . 'INFO' . $middle . 'Info' . $end;

            default:
                return '';
        }
    }
}


src/Style.php
<?php

declare(strict_types=1);

namespace BlueConsole;

use Symfony\Component\Console\Style\SymfonyStyle;

class Style extends SymfonyStyle
{
    use AdditionalStyles;

    public const DAY_SECONDS = 86400;

    /**
     * @param bool $newLine
     * @return Style
     * @throws \Exception
     */
    public function timer(bool $newLine = true): self
    {
        $this->write($this->getTimer());

        if ($newLine) {
            $this->newLine();
        }

        return $this;
    }

    /**
     * @param string $section
     * @param string $message
     * @param string $style
     * @return $this
     * @throws \Exception
     * @throws \InvalidArgumentException
     */
    public function formatSection(string $section, string $message, string $style = 'info'): self
    {
        $timer = $this->getTimer(true);

        $this->writeln(
            $timer . $this->formatter->formatSection(
                $section,
                $message,
                $style
            )
        );

        return $this;
    }

    /**
     * @param string|array $messages
     * @param string $style
     * @param bool $large
     * @return $this
     * @throws \Exception
     * @throws \InvalidArgumentException
     * @noinspection MissingParameterTypeDeclarationInspection
     */
    public function formatBlock($messages, string $style, bool $large = false): self
    {
        $timer = $this->getTimer(true);

        if (!\is_array($messages)) {
            $messages = [$messages];
        }

        foreach ($messages as $message) {
            $this->writeln(
                $timer . $this->formatter->formatBlock(
                    $message,
                    $style,
                    $large
                )
            );
        }

        return $this;
    }

    /**
     * @param array $message
     * @return $this
     * @throws \Exception
     * @throws \InvalidArgumentException
     */
    public function errorLine(array $message): self
    {
        return $this->formatBlock($message, 'error');
    }

    /**
     * @param $message
     * @return $this
     * @throws \InvalidArgumentException
     * @throws \Exception
     */
    public function okMessage(string $message): self
    {
        return $this->renderBlock('  <info>OK</info>  ', $message);
    }

    /**
     * @param $message
     * @return $this
     * @throws \InvalidArgumentException
     * @throws \Exception
     */
    public function errorMessage(string $message): self
    {
        return $this->renderBlock(' <fg=red>FAIL</> ', $message);
    }

    /**
     * @param string $message
     * @return $this
     * @throws \InvalidArgumentException
     * @throws \Exception
     */
    public function warningMessage(string $message): self
    {
        return $this->renderBlock(' <comment>WARN</comment> ', $message);
    }

    /**
     * @param string $message
     * @return $this
     * @throws \InvalidArgumentException
     * @throws \Exception
     */
    public function infoMessage(string $message): self
    {
        return $this->renderBlock(' <fg=blue>INFO</> ', $message);
    }

    /**
     * @param string $block
     * @param string $message
     * @return Style
     * @throws \Exception
     */
    public function renderBlock(string $block, string $message): self
    {
        $timer = $this->getTimer(true);
        $alignment = $this->align(8, $this->align);

        $this->write("{$timer}[$block]");
        $this->write($alignment);
        $this->writeln($message);

        return $this;
    }

    /**
     * @param string $message
     * @return $this
     * @throws \InvalidArgumentException
     * @throws \Exception
     */
    public function note($message): self
    {
        return $this->genericBlock($message, 'blue', 'note');
    }

    /**
     * @param string $message
     * @return $this
     * @throws \InvalidArgumentException
     * @throws \Exception
     */
    public function caution($message): self
    {
        return $this->genericBlock($message, 'magenta', 'caution');
    }

    /**
     * @param string $message
     * @return $this
     * @throws \InvalidArgumentException
     * @throws \Exception
     */
    public function success($message): self
    {
        return $this->genericBlock($message, 'green', 'success');
    }

    /**
     * @param string $message
     * @return $this
     * @throws \InvalidArgumentException
     * @throws \Exception
     */
    public function warning($message): self
    {
        return $this->genericBlock($message, 'yellow', 'warning');
    }

    /**
     * @param string $message
     * @return $this
     * @throws \InvalidArgumentException
     * @throws \Exception
     */
    public function error($message): self
    {
        return $this->genericBlock($message, 'red', 'error');
    }

    /**
     * @param string $message
     * @param string $background
     * @param string|null $type
     * @param int $length
     * @return $this
     * @throws \InvalidArgumentException
     * @throws \Exception
     */
    public function genericBlock(string $message, string $background, ?string $type, int $length = 100): self
    {
        if (\trim($type) === '') {
            return $this;
        }

        $type = \strtoupper($type);
        $alignment = $this->align(0, $length);
        $alignmentMessage = $this->align($message, $length - (\mb_strlen($type) + 5));

        $timer = $this->getTimer(true);
        $timer ? $this->writeln($timer) : null;

        $this->writeln("<bg=$background>$alignment</>");
        $this->writeln("<fg=white;bg=$background>  [$type] $message$alignmentMessage</>");
        $this->writeln("<bg=$background>$alignment</>");
        $this->newLine();

        return $this;
    }

    public function floatingProgressBar(): void
    {
        
    }

    public function longRunCommand(string $command): void
    {
        //[ ***  ]
    }

    /**
     * @todo add multi line block
     */

    /*
        http://symfony.com/doc/current/components/console/helpers/formatterhelper.html
        https://symfony.com/doc/current/console/style.html
    */
}


src/Animation.php
<?php

namespace BlueConsole;

class Animation
{
    /**
     * @var \DateInterval
     */
    protected $interval;

    public function __construct(int $steps = 6, string $cursor = '*')
    {
        $counter = 0;
        $forward = true;
        $before = 0;
        $after = $steps - \strlen($cursor);

        echo "\n";

        while (true) {
            echo '[';

            for ($i = 1; $i <= $before; $i++) {
                echo ' ';
            }

            echo $cursor;

            for ($i = 0; $i <= $after; $i++) {
                echo ' ';
            }

            echo ']';
            echo "\r";

            $counter++;

            if ($forward) {
                $before++;
                $after--;
            } else {
                $before--;
                $after++;
            }

            if ($counter === ($steps - (\strlen($cursor) -1))) {
                $counter = 0;
                $forward = !$forward;
            }

            sleep(1);
        }
    }

    public function setWaitTime(\DateInterval $dateInterval)
    {
        //        $interval = new \DateInterval('PT10S');
        $dateInterval->invert = true;
        //dump($interval);

        $d1 = new \DateTime();
        while (true) {
            $d2 = new \DateTime();
            $d2->add($dateInterval);

            $iv = $d2->diff($d1);
            dump($iv->s);
            if ($iv->s === 0) {
                echo "end\n";
                exit(0);
            }
            sleep(1);
        }
    }

    public function setMessage(string $message)
    {
        
    }

    public function run($process)
    {
        //display message with animation
        //execute external process
        //wait and display animation
        //process ended -> display message and go fodder
    }

    protected function displayAnimation()
    {
        $counter = 0;
        $forward = true;
        $before = 0;
        $after = $steps - \strlen($cursor);

        while (true) {
            echo '[';

            for ($i = 1; $i <= $before; $i++) {
                echo ' ';
            }

            echo $cursor;

            for ($i = 0; $i <= $after; $i++) {
                echo ' ';
            }

            echo ']';
            echo "\r";

            $counter++;

            if ($forward) {
                $before++;
                $after--;
            } else {
                $before--;
                $after++;
            }

            if ($counter === ($steps - (\strlen($cursor) -1))) {
                $counter = 0;
                $forward = !$forward;
            }

            sleep(1);
        }
    }

    protected function executeProcess()
    {
        
    }
}


examples/style.php
#!/usr/bin/env php
<?php

namespace SymfonyStyle\Examples;

require_once __DIR__ . '/../vendor/autoload.php';

use BlueConsole\Style;
use Symfony\Component\Console\{
    Input\ArgvInput,
    Output\ConsoleOutput,
    Helper\FormatterHelper,
};

$input = new ArgvInput;
$output = new ConsoleOutput;
$formatter = new FormatterHelper;
$style = new Style($input, $output, $formatter);

$style->warningMessage('Warning message');
$style->errorMessage('Error message');
$style->okMessage('Ok message');
$style->infoMessage('Info message');

$style->error('Error');
$style->note('Note');
$style->warning('Warning');
$style->caution('Caution');
$style->success('Success');

$style->genericBlock('Generic block success', 'blue', 'success', 50);
$style->genericBlock('Generic block success', 'blue', '', 50);

$style->formatBlock(['Error 1', 'Error 2'], 'error');
$style->formatBlock(['Error 1', 'Error 2'], 'info');
$style->errorLine(['Error Line']);

$style->formatSection('Section', 'Format Section info (default)');
$style->formatSection('Section', 'Format Section error', 'error');
$style->formatSection('Section', 'Format Section comment', 'comment');
$style->formatSection('Section', 'Format Section question', 'question');

$style->setAlign(5);
$style->infoMessage('Alignment 5');
$style->setAlign(10);
$style->okMessage('Alignment 10');
$style->setAlign(15);
$style->warningMessage('Alignment 15');

$message = 'This is a very long message, which should be truncated';
$style->truncateLn($message, 7);
$style->truncateLn($message, -5);
$style->truncateLn($message, 7, '!!');
$style->truncateLn('Short', 10);

//timer display enable
$style->newLine(4);

$style->toggleShowTimer();

$style->timer();

$style->warningMessage('Warning message with timer');
$style->errorMessage('Error message with timer');
$style->okMessage('Ok message with timer');
$style->infoMessage('Info message with timer');

$style->newLine(2);
$style->error('Error with timer');
$style->note('Note with timer');
$style->warning('Warning with timer');
$style->caution('Caution with timer');
$style->success('Success with timer');

$style->genericBlock('Generic block success', 'blue', 'success', 50);
$style->genericBlock('Generic block success', 'blue', '', 50);

$style->formatBlock(['Error 1', 'Error 2'], 'error');
$style->formatBlock(['Error 1', 'Error 2'], 'info');
$style->errorLine(['Error Line']);

$style->formatSection('Section', 'Format Section info (default)');
$style->formatSection('Section', 'Format Section error', 'error');
$style->formatSection('Section', 'Format Section comment', 'comment');
$style->formatSection('Section', 'Format Section question', 'question');

$style->setAlign(5);
$style->infoMessage('Alignment 5');
$style->setAlign(10);
$style->okMessage('Alignment 10');
$style->setAlign(15);
$style->warningMessage('Alignment 15');

//display datetime
$style->newLine(4);

$style->toggleTimerType();

$style->warningMessage('Warning message with timer');
$style->errorMessage('Error message with timer');
$style->okMessage('Ok message with timer');
$style->infoMessage('Info message with timer');
